### PR TITLE
Misc additions

### PR DESCRIPTION
**Pull request by DeNeutoy**
Here were some things that are actually unrelated to the `Trainer` PR.

- Changes the `read()` signature of `DatasetReaders` to `read(filepath)`
- Add a function which converts the output of a Dataset's `as_arrays` method into something suitable for consumption by a `Model`.
- Change the `ensure_model_trains_and_loads` function for testing to `ensure_model_saves_and_loads` as a preliminary step towards using the `Trainer` and `Driver` classes in testing models.
- Adds a function to create an `optimizer` from json and the model parameters to be optimised.

### CODE REVIEW

**Comment by DeNeutoy** [comment]
@joelgrus - I not sure mypy [supports recursive typing](https://stackoverflow.com/questions/44549824/invalid-type-error-in-self-referential-mypy-types
) (re: our discussion yesterday).

I can ignore that one, but I don't understand why the changes I have made to the dataset readers have broken the `TokenIndexer` types. Can you explain when you get in?

**Comment by joelgrus** [comment]
that sucks about the recursive types. two options:

(1) if you know you only want (say) one level deep, you could do 

```
DictOfArrays = Dict[str, Union[Dict[str, Union[Dict, np.array]], numpy.array]]
```

(2) if you want to mostly fake it, you could do

```
DictOfArrays = Dict[str, union[dict, numpy.array]]
```

I'll look at the other issue

File: allennlp/common/tensor.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+from typing import Dict, Optional, Union
+
+import torch
+from torch.autograd import Variable
+import numpy
+
+DictOfArrays = Dict[str, Union['DictOfArrays', numpy.array]]  # pylint: disable=invalid-name
```
Comment (joelgrus) [inline allennlp/common/tensor.py]: `numpy.array` is a function, `numpy.ndarray` is the data type

Comment (matt-gardner) [inline allennlp/common/tensor.py]: Oops...  I had that all over the place in `DeepQA`.  Useful for documentation and readability, but clearly not correctly typed.

**Comment by joelgrus** [comment]
still not 100% sure why the other typing issue popped up, but you can fix it by adding a couple of type annotations:

squad.py:205

```
sentence_fields = []  # type: List[Field]
```

language_modeling.py:73

```
output_indexer = None  # type: Dict[str, TokenIndexer]
```

**Comment by DeNeutoy** [comment]
@matt-gardner , @joelgrus this is ready for another look.

File: allennlp/common/tensor.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        for key, value in data_structure.items():
+            data_structure[key] = data_structure_as_variables(value)
+        return data_structure
+    else:
+
+        torch_variable = Variable(torch.from_numpy(data_structure))
+        if cuda_device == -1:
+            return torch_variable
+        else:
+            return torch_variable.cuda(cuda_device)
```
Comment (DeNeutoy) [inline allennlp/common/tensor.py]: This is true, but only if you have a GPU version of pytorch installed (otherwise calling any `.cuda` stuff raises compilation errors). This allows this function to be used cross platform.

File: allennlp/common/tensor.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+from typing import Dict, Union
+
+import torch
+from torch.autograd import Variable
+import numpy
+
+
+def data_structure_as_variables(data_structure: Dict[str, Union[dict, numpy.ndarray]],
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: `arrays_to_variables`?  Seems like we can find a simpler name here...

File: allennlp/common/tensor.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+                                cuda_device: int = -1):
+
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: This blank line doesn't add anything.

File: allennlp/common/tensor.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    if isinstance(data_structure, dict):
+        for key, value in data_structure.items():
+            data_structure[key] = data_structure_as_variables(value)
+        return data_structure
+    else:
+
```
Comment (matt-gardner) [inline allennlp/common/tensor.py]: Remove blank line.

File: allennlp/testing/test_case.py [hunk 6]
```diff
@@ -33,63 +40,47 @@ def setUp(self):
...
-            for model_prediction, loaded_prediction in zip(model_predictions, loaded_model_predictions):
-                assert_allclose(model_prediction, loaded_prediction)
+    def ensure_model_saves_and_loads(self,
+                                     model: Model,
+                                     dataset_reader: DatasetReader,
+                                     iterator: DataIterator = None):
+
+        # Our loading tests work better if you're not using complex iterators, so by
+        # default we use the basic one unless you pass an iterator into this function.
+        # If you _do_ use them, we'll skip some of the stuff below that isn't compatible.
```
Comment (matt-gardner) [inline allennlp/testing/test_case.py]: I think the issue here was with shuffling, which is now easy to disable.  But you don't have any checks for them anymore, anyway, so this comment can be removed.

File: allennlp/testing/test_case.py [hunk 7]
```diff
@@ -33,63 +40,47 @@ def setUp(self):
...
+        data_iterator = iterator or BasicIterator()
+
+        dataset = dataset_reader.read(self.TRAIN_FILE)
+        vocab = Vocabulary.from_dataset(dataset)
+        dataset.index_instances(vocab)
+
+        single_batch = next(data_iterator(dataset))
+        print(single_batch)
```
Comment (matt-gardner) [inline allennlp/testing/test_case.py]: Remove print statement.

File: allennlp/training/optimizers.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        "adagrad": torch.optim.Adagrad,
+        "adadelta": torch.optim.Adadelta,
+        "sgd": torch.optim.SGD,
+        "rmsprop": torch.optim.RMSprop,
+
+}
+
+
+def get_optimizer_from_params(model_parameters: List[torch.Tensor],
+                              params: Union[Dict[str, Any], str]) -> torch.optim.Optimizer:
```
Comment (matt-gardner) [inline allennlp/training/optimizers.py]: Should this be a `Params` object, or at least `Union[Params, str]`?

File: allennlp/training/optimizers.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    This function converts a set of model parameters to be optimized and a
+    dictionary of optimizer parameters (or a string name of one of the optimizers above)
+    into a Pytorch optimizer object which can apply gradient updates to the model
+    parameters following model.backward() calls.
+    The simplest case for both of these is a string that shows up in `optimizers`
+    above - if `params` is just one of those strings, we return it, and everyone
+    is happy. If not, we assume `params` is a Dict[str, Any], with a "type" key,
+    where the value for "type" must be one of those strings above. We take the rest
+    of the parameters and pass them to the optimizer's constructor.
+
```
Comment (matt-gardner) [inline allennlp/training/optimizers.py]: Remove blank line.

File: allennlp/training/optimizers.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+def get_optimizer_from_params(model_parameters: List[torch.Tensor],
+                              params: Union[Dict[str, Any], str]) -> torch.optim.Optimizer:
+    """
+    This function converts a set of model parameters to be optimized and a
+    dictionary of optimizer parameters (or a string name of one of the optimizers above)
+    into a Pytorch optimizer object which can apply gradient updates to the model
+    parameters following model.backward() calls.
+    The simplest case for both of these is a string that shows up in `optimizers`
```
Comment (matt-gardner) [inline allennlp/training/optimizers.py]: You need double back ticks for code (throughout this docstring), single back tick is italics.

File: allennlp/training/optimizers.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+    is happy. If not, we assume `params` is a Dict[str, Any], with a "type" key,
+    where the value for "type" must be one of those strings above. We take the rest
+    of the parameters and pass them to the optimizer's constructor.
+
+    """
+    if isinstance(params, str):
+        optimizer = params
+        params = {}
+    else:
+        optimizer = params.pop("type")
```
Comment (matt-gardner) [inline allennlp/training/optimizers.py]: `params.pop_choice`?

File: allennlp/training/optimizers.py (hunk 8 above)
Comment (DeNeutoy) [inline allennlp/training/optimizers.py]: I changed this as I thought we might need to pass these optimizer parameters to the `Trainer` constructor as dictionary arguments because in order to instantiate the optimizer, we need the parameters of the model, which I couldn't see how to do in a `from_params` call pleasantly. I changed this back though, as maybe there is a way.
